### DISCUSSION TITLE
Best way to run a migration on first persist?

### DISCUSSION THREAD

**Discussion by pvinis** [pmndrs/zustand#1717] [Q&A]
- So I have an app that used raw AsyncStorage.
- Now I want to put zustand with persist.
- I thought I can use `version: 1` and then
- in the `migrate` function to have v1 do a migration from AsyncStorage to zustand.

I can't find a good way to do this. zustand state goes to v1 directly with no migration, since there is no persisted state already in v0.

any way I could do that better? or a way to force zustand to run the migration?

Thanks.

**Comment by JacobWeisenburger** [comment]
zustand runs a migration anytime there is a version mismatch between the version in `PersistOptions` and the version in `localStorage`. So you just need to put a version into `PersistOptions` and it will run the migration function.

```ts
const store = createStore(
    persist(() => ({}), {
        name: 'storage-key',
        version: 1,
        migrate(prevState, prevVersion) {
            console.log({ prevState, prevVersion })
            // 1st time: doesn't run
            // 2nd time: { prevState: {}, prevVersion: 0 }
            return { foo: 'bar' }
        },
    })
)

console.log(localStorage.getItem('storage-key'))
// 1st time: {"state":{},"version":0}
// 2nd time: {"state":{"foo":"bar"},"version":1}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by atothewest** [comment]
This question is quite old now but for anyone stumbling upon it I am not sure the answer given above fully answers the question (with the latest version of Zustand at least). The `migrate` function is only run [if a previously persisted version is found in storage](https://github.com/pmndrs/zustand/blob/a7b57b24d5fbffaeed3c2078efa39fbe7e4dd756/src/middleware/persist.ts#L451), so on the first render with empty storage, it is skipped and the store is initialised with whatever the version is specified as – as the OP suggests.

It would be useful IMO to be able to perform an "empty" migration, but a workaround is possible using `merge` instead. This function will be called with the value of any persisted state along with the current (initial) state, but it's also possible to inject values at this point also. For example:

```
const store = createStore<T>()(
  persist(() => ({}),
  {
    name: 'my-store',
    merge: (persistedState, currentState) => {
      if (!persistedState) {
        // storage is empty, perform migration
        const stateToMigrate = someLegacyStateFetch();
        return { ...currentState, ...stateToMigrate };
      }

     return { ...currentState, ...persistedState };
    }
  }
);
```

**Reply by kvndrsslr** [reply]
For anyone looking into this: note that the merge function is not called when the underlying store experiences an exception while getting the item (if you previously stored something that is not valid JSON).

**Comment by albertusdev** [comment]
I tried out both solution above but was not able to find a way to force the migrate function to run on initial persist, until I read the source code of the `migrate` implementation [here](https://github.com/pmndrs/zustand/blob/main/src/middleware/persist.ts#L302) and the precondition for it to run.

> The migrate function is only run [if a previously persisted version is found in storage](https://github.com/pmndrs/zustand/blob/a7b57b24d5fbffaeed3c2078efa39fbe7e4dd756/src/middleware/persist.ts#L451), so on the first render with empty storage, it is skipped and the store is initialised with whatever the version is specified as – as the OP suggests.

@atothewest is correct here that the migration will be skipped if the storage is empty since the migration will only run if the previous `getItem` returned a number.

A workaround to always force migration to run on initial persist would be to return `{ version: 0 }` as part of the `getItem` in the storage.

To make it simpler, I have created a decorator below that can be used as a workaround to run the migration on first persist:

```ts
import { create } from 'zustand';
import { createJSONStorage, persist, PersistStorage, type StateStorage, StorageValue } from 'zustand/middleware';

export function forceMigrationOnInitialPersist<S>(originalStorage: PersistStorage<S> | undefined): PersistStorage<S> | undefined {
  try {
    if (!originalStorage) {
      return undefined;
    }
    const modifiedStorage: PersistStorage<S> = {
      getItem: async name => {
        const item = await originalStorage.getItem(name);
        if (item) {
          return item;
        }
        return {
          state: undefined,
          version: 0
        } as StorageValue<S>;
      },
      setItem: originalStorage.setItem,
      removeItem: originalStorage.removeItem
    };
    return modifiedStorage;
  } catch (e) {
    console.error('forceMigrationOnInitialPersist', e);
    return undefined;
  }
}

export const useAuthStore = create(
  persist(
    () => ({
      foo: 'bar',
    }),
    {
      name: 'auth-storage',
      storage: forceMigrationOnInitialPersist(createJSONStorage(() => localStorage)),
      version: 1,
      migrate: async (state, version) => {
        if (version === 0) {
          // run your initial/forced migration here! it supports async, so you can read IndexedDB or anything similar here
          // the state would be the default state here, in ths case: { foo: 'bar' }
        } 
        return state;
      }
    }
  )
);
```
